src/impl_/concat.rs

```diff
@@ -6,25 +6,210 @@ macro_rules! const_concat {
         $e
     }};
     ($l:expr, $($r:expr),+ $(,)?) => {{
-        const L: &'static str = $l;
-        const R: &'static str = $crate::impl_::concat::const_concat!($($r),*);
-        const LEN: usize = L.len() + R.len();
-        const fn combine(l: &'static [u8], r: &'static [u8]) -> [u8; LEN] {
-            let mut out = [0u8; LEN];
-            let mut i = 0;
-            while i < l.len() {
-                out[i] = l[i];
-                i += 1;
+        const PIECES: &[&str] = &[$l, $($r),+];
+        const RAW_BYTES: &[u8] = &$crate::impl_::concat::combine_to_array::<{
+            $crate::impl_::concat::combined_len(PIECES)
+        }>(PIECES);
+        // Safety: `RAW_BYTES` is combined from valid &str pieces
+        unsafe { ::std::str::from_utf8_unchecked(RAW_BYTES) }
+    }}
+}
+
+pub use const_concat;
+
+/// Calculates the total byte length of all string pieces in the array.
+///
+/// This is a useful utility in order to determine the size needed for the constant
+/// `combine` function.
+pub const fn combined_len(pieces: &[&str]) -> usize {
+    let mut len = 0;
+    let mut pieces_idx = 0;
+    while pieces_idx < pieces.len() {
+        len += pieces[pieces_idx].len();
+        pieces_idx += 1;
+    }
+    len
+}
+
+/// Combines all string pieces into a single byte array.
+///
+/// `out` should be a buffer at the correct size of `combined_len(pieces)`, else this will panic.
+#[cfg(mut_ref_in_const_fn)] // requires MSRV 1.83
+pub const fn combine(pieces: &[&str], out: &mut [u8]) {
+    let mut out_idx = 0;
+    let mut pieces_idx = 0;
+    while pieces_idx < pieces.len() {
+        let piece = pieces[pieces_idx].as_bytes();
+        let mut piece_idx = 0;
+        while piece_idx < piece.len() {
+            out[out_idx] = piece[piece_idx];
+            out_idx += 1;
+            piece_idx += 1;
+        }
```

Comment:
This could be `copy_from_slice`, but would need another cfg for MSRV. Not sure if worth it, maybe a comment is enough?

Reply:
Good point - I pushed a commit which factors this out into a common helper and leaves a suitable comment on top of it.